src/datachain/lib/dc.py

```diff
@@ -438,6 +438,18 @@ def from_storage(
             uri, session, update=update
         )
 
+        # ds_name is None if object is a file, we don't want to use cache
+        # or do listing in that case - just read that single object
+        if not list_ds_name:
```

Comment:
[C]: main change here (along with `get_listing` that now returns empty ds_name in case of a single file). Everything else is pretty much fixing things to work with this change.